File: internal/embedding/worker.go
package embedding

import (
	"context"
	"log/slog"
	"sync"
	"time"

	"argus/internal/model"
	"argus/internal/store"
)

// Worker runs a background loop that embeds unembedded messages/chunks
// and generates summaries for long assistant replies.
type Worker struct {
	client       *Client
	semantic     store.SemanticStore
	pinned       store.PinnedMemoryStore
	docs         store.DocumentStore
	summarizer   model.Client // cheap/fast model for summary generation (may be nil)
	batchSize    int
	interval     time.Duration
	notify       chan struct{}
	quit         chan struct{}
	wg           sync.WaitGroup
}

func NewWorker(client *Client, semantic store.SemanticStore, pinned store.PinnedMemoryStore, docs store.DocumentStore, batchSize int, interval time.Duration) *Worker {
	if batchSize == 0 {
		batchSize = 32
	}
	if interval == 0 {
		interval = 2 * time.Second
	}
	return &Worker{
		client:    client,
		semantic:  semantic,
		pinned:    pinned,
		docs:      docs,
		batchSize: batchSize,
		interval:  interval,
		notify:    make(chan struct{}, 1),
		quit:      make(chan struct{}),
	}
}

func (w *Worker) Start() {
	w.wg.Add(1)
	go w.run()
	slog.Info("embedding worker started", "batch_size", w.batchSize, "interval", w.interval)
}

func (w *Worker) Stop() {
	close(w.quit)
	w.wg.Wait()
	slog.Info("embedding worker stopped")
}

// Notify signals the worker to wake up and process new content.
func (w *Worker) Notify() {
	select {
	case w.notify <- struct{}{}:
	default: // already notified
	}
}

func (w *Worker) run() {
	defer w.wg.Done()

	ticker := time.NewTicker(w.interval)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			w.processAll()
		case <-w.notify:
			w.processAll()
		case <-w.quit:
			return
		}
	}
}

// SetSummarizer configures the model client used for async summary generation.
func (w *Worker) SetSummarizer(client model.Client) {
	w.summarizer = client
}

func (w *Worker) processAll() {
	ctx := context.Background()

	// Embed unembedded messages.
	if w.semantic != nil {
		w.embedMessages(ctx)
	}

	// Embed unembedded chunks.
	if w.docs != nil {
		w.embedChunks(ctx)
	}

	// Summarize long assistant replies.
	if w.summarizer != nil && w.semantic != nil {
		w.summarizeMessages(ctx)
	}
}

func (w *Worker) embedMessages(ctx context.Context) {
	msgs, err := w.semantic.UnembeddedMessages(ctx, w.batchSize)
	if err != nil {
		slog.Warn("fetch unembedded messages", "err", err)
		return
	}
	if len(msgs) == 0 {
		return
	}

	texts := make([]string, len(msgs))
	for i, m := range msgs {
		texts[i] = m.Content
	}

	vectors, err := w.client.Embed(ctx, texts)
	if err != nil {
		// Batch failed — fallback to one-by-one to isolate the bad message.
		slog.Warn("embed messages batch failed, trying individually", "err", err, "count", len(texts))
		for i, m := range msgs {
			vec, err := w.client.Embed(ctx, []string{texts[i]})
			if err != nil {
				slog.Warn("embed single message failed", "id", m.ID, "err", err)
				continue
			}
			if len(vec) > 0 && vec[0] != nil {
				w.semantic.SetMessageEmbedding(ctx, m.ID, vec[0])
			}
		}
		return
	}

	for i, vec := range vectors {
		if vec == nil {
			continue
		}
		if err := w.semantic.SetMessageEmbedding(ctx, msgs[i].ID, vec); err != nil {
			slog.Warn("set message embedding", "id", msgs[i].ID, "err", err)
		}
	}

	slog.Info("embedded messages", "count", len(msgs))
}

func (w *Worker) embedChunks(ctx context.Context) {
	chunks, err := w.docs.UnembeddedChunks(ctx, w.batchSize)
	if err != nil {
		slog.Warn("fetch unembedded chunks", "err", err)
		return
	}
	if len(chunks) == 0 {
		return
	}

	texts := make([]string, len(chunks))
	for i, c := range chunks {
		texts[i] = c.Content
	}

	vectors, err := w.client.Embed(ctx, texts)
	if err != nil {
		slog.Warn("embed chunks batch failed, trying individually", "err", err, "count", len(texts))
		for i, c := range chunks {
			vec, err := w.client.Embed(ctx, []string{texts[i]})
			if err != nil {
				slog.Warn("embed single chunk failed", "id", c.ID, "err", err)
				continue
			}
			if len(vec) > 0 && vec[0] != nil {
				w.docs.SetChunkEmbedding(ctx, c.ID, vec[0])
			}
		}
		return
	}

	for i, vec := range vectors {
		if vec == nil {
			continue
		}
		if err := w.docs.SetChunkEmbedding(ctx, chunks[i].ID, vec); err != nil {
			slog.Warn("set chunk embedding", "id", chunks[i].ID, "err", err)
		}
	}

	slog.Info("embedded chunks", "count", len(chunks))
}

const summaryPrompt = `Summarize the following assistant reply in 2-3 concise sentences.
Preserve the key facts, conclusions, and any specific data points.
Use the same language as the original. Output ONLY the summary, nothing else.`

func (w *Worker) summarizeMessages(ctx context.Context) {
	msgs, err := w.semantic.UnsummarizedMessages(ctx, 5)
	if err != nil {
		slog.Warn("fetch unsummarized messages", "err", err)
		return
	}
	if len(msgs) == 0 {
		return
	}

	for _, m := range msgs {
		messages := []model.Message{
			{Role: model.RoleSystem, Content: summaryPrompt},
			{Role: model.RoleUser, Content: m.Content},
		}
		resp, err := w.summarizer.Chat(ctx, messages, nil)
		if err != nil {
			slog.Warn("summarize message failed, will retry next cycle", "id", m.ID, "err", err)
			continue
		}
		if err := w.semantic.SetMessageSummary(ctx, m.ID, resp.Content); err != nil {
			slog.Warn("set message summary", "id", m.ID, "err", err)
		}
	}

	slog.Info("summarized messages", "count", len(msgs))
}

File: src/agent/worker.rs
//! Background workers: embedder and summarizer.
//! Spawned by Agent, run until cancelled.

use std::sync::Arc;
use std::time::Duration;

use tokio_util::sync::CancellationToken;
use tracing::{debug, info, warn};

use super::embedding::EmbeddingClient;
use crate::database::Database;
use crate::upstream;
use crate::upstream::types as model;

const SUMMARY_PROMPT: &str = "Summarize the following assistant reply in 2-3 concise sentences. \
Preserve the key facts, conclusions, and any specific data points. \
Use the same language as the original. Output ONLY the summary, nothing else.";

/// Spawn the embedder background task. Returns its join handle.
pub(super) fn spawn_embedder(
    db: Arc<Database>,
    embedder: Arc<EmbeddingClient>,
    cancel: CancellationToken,
    batch_size: usize,
    interval: Duration,
) -> tokio::task::JoinHandle<()> {
    tokio::spawn(async move {
        info!("embedder worker started");
        loop {
            tokio::select! {
                _ = tokio::time::sleep(interval) => {
                    embed_cycle(&db, &embedder, batch_size).await;
                }
                _ = cancel.cancelled() => break,
            }
        }
        info!("embedder worker stopped");
    })
}

/// Spawn the summarizer background task. Returns its join handle.
pub(super) fn spawn_summarizer(
    db: Arc<Database>,
    synthesizer: Arc<dyn upstream::Client>,
    cancel: CancellationToken,
    interval: Duration,
) -> tokio::task::JoinHandle<()> {
    tokio::spawn(async move {
        info!("summarizer worker started");
        loop {
            tokio::select! {
                _ = tokio::time::sleep(interval) => {
                    summarize_cycle(&db, &*synthesizer).await;
                }
                _ = cancel.cancelled() => break,
            }
        }
        info!("summarizer worker stopped");
    })
}

/// Spawn the document ingester background task.
pub(super) fn spawn_ingester(
    db: Arc<Database>,
    cancel: CancellationToken,
    interval: Duration,
) -> tokio::task::JoinHandle<()> {
    tokio::spawn(async move {
        info!("ingester worker started");
        loop {
            tokio::select! {
                _ = tokio::time::sleep(interval) => {
                    ingest_cycle(&db).await;
                }
                _ = cancel.cancelled() => break,
            }
        }
        info!("ingester worker stopped");
    })
}

/// One ingestion cycle: process pending large documents.
async fn ingest_cycle(db: &Database) {
    let docs = match db.documents.pending(5).await {
        Ok(d) => d,
        Err(e) => {
            warn!(error = %e, "fetch pending documents failed");
            return;
        }
    };
    for doc in &docs {
        if let Err(e) = super::docindex::ingest_document(
            db,
            doc.id,
            &doc.filename,
            std::path::Path::new(&doc.file_path),
        )
        .await
        {
            warn!(doc_id = doc.id, filename = doc.filename, error = %e, "document ingestion failed");
            let _ = db
                .documents
                .update_status(doc.id, "error", &e.to_string())
                .await;
        }
    }
}

/// One embedding cycle: fetch unembedded messages + notifications + chunks, embed, write back.
async fn embed_cycle(db: &Database, embedder: &EmbeddingClient, batch_size: usize) {
    embed_table(&db.messages, embedder, batch_size, "messages").await;
    embed_table(&db.notifications, embedder, batch_size, "notifications").await;
    embed_table(&db.documents, embedder, batch_size, "chunks").await;
}

/// Generic embed helper for any table that has unembedded() + set_embedding().
async fn embed_table(
    table: &(impl Embeddable + Sync),
    embedder: &EmbeddingClient,
    batch_size: usize,
    label: &str,
) {
    let rows = match table.unembedded(batch_size as i64).await {
        Ok(r) if !r.is_empty() => r,
        _ => return,
    };

    let texts: Vec<&str> = rows.iter().map(|(_, c)| c.as_str()).collect();
    match embedder.embed_batch(&texts).await {
        Ok(vecs) => {
            for ((id, _), vec) in rows.iter().zip(vecs.iter()) {
                if let Err(e) = table.set_embedding(*id, vec).await {
                    warn!(id, label, error = %e, "set embedding failed");
                }
            }
            debug!(count = rows.len(), label, "embedded");
        }
        Err(e) => warn!(label, error = %e, "batch embed failed"),
    }
}

/// Trait for tables that support embedding.
#[async_trait::async_trait]
trait Embeddable {
    async fn unembedded(&self, limit: i64) -> anyhow::Result<Vec<(i64, String)>>;
    async fn set_embedding(&self, id: i64, embedding: &[f32]) -> anyhow::Result<()>;
}

#[async_trait::async_trait]
impl Embeddable for crate::database::Messages {
    async fn unembedded(&self, limit: i64) -> anyhow::Result<Vec<(i64, String)>> {
        self.unembedded(limit).await
    }
    async fn set_embedding(&self, id: i64, embedding: &[f32]) -> anyhow::Result<()> {
        self.set_embedding(id, embedding).await
    }
}

#[async_trait::async_trait]
impl Embeddable for crate::database::Documents {
    async fn unembedded(&self, limit: i64) -> anyhow::Result<Vec<(i64, String)>> {
        self.unembedded_chunks(limit).await
    }
    async fn set_embedding(&self, id: i64, embedding: &[f32]) -> anyhow::Result<()> {
        self.set_chunk_embedding(id, embedding).await
    }
}

#[async_trait::async_trait]
impl Embeddable for crate::database::Notifications {
    async fn unembedded(&self, limit: i64) -> anyhow::Result<Vec<(i64, String)>> {
        self.unembedded(limit).await
    }
    async fn set_embedding(&self, id: i64, embedding: &[f32]) -> anyhow::Result<()> {
        self.set_embedding(id, embedding).await
    }
}

/// One summarization cycle: fetch long unsummarized notifications, summarize, write back.
async fn summarize_cycle(db: &Database, synthesizer: &dyn upstream::Client) {
    let rows = match db.notifications.unsummarized(5).await {
        Ok(r) => r,
        Err(e) => {
            warn!(error = %e, "fetch unsummarized failed");
            return;
        }
    };

    for (id, content) in &rows {
        let messages = vec![
            model::Message::system(SUMMARY_PROMPT),
            model::Message::user(content.as_str()),
        ];
        match synthesizer.chat(&messages, &[]).await {
            Ok(resp) => {
                if let Err(e) = db.notifications.set_summary(*id, &resp.content).await {
                    warn!(id, error = %e, "set summary failed");
                }
            }
            Err(e) => {
                warn!(id, error = %e, "summarize failed, will retry");
            }
        }
    }

    if !rows.is_empty() {
        debug!(count = rows.len(), "notifications summarized");
    }
}
